llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp

```diff
@@ -5720,6 +5737,23 @@ SDValue DAGTypeLegalizer::WidenVecRes_VP_STRIDED_LOAD(VPStridedLoadSDNode *N) {
   return Res;
 }
 
+SDValue DAGTypeLegalizer::WidenVecRes_MASKED_COMPRESS(SDNode *N) {
+  SDValue Vec = N->getOperand(0);
+  SDValue Mask = N->getOperand(1);
+  SDValue Passthru = N->getOperand(2);
+  EVT WideVecVT =
+      TLI.getTypeToTransformTo(*DAG.getContext(), Vec.getValueType());
+  EVT WideMaskVT = EVT::getVectorVT(*DAG.getContext(),
+                                    Mask.getValueType().getVectorElementType(),
+                                    WideVecVT.getVectorNumElements());
+
+  SDValue WideVec = ModifyToType(Vec, WideVecVT);
+  SDValue WideMask = ModifyToType(Mask, WideMaskVT);
```

Comment:
Do you need to pass FillWithZeros=true for the mask?

Reply:
In my understanding, we don't. If we have a `<3 x i32>` but need a `<4 x i32>`, the "new" fourth element is only relevant for the register but not to the user. The fourth `Vec` and `Passthru` values are both `undef`, so it does not really matter if the mask selects them or not. Please correct me if I'm wrong.

Reply:
Suppose you have a mask of `<false, false, true>`, and a passthru.  If you widen the mask to `<false, false, true, true>`, the second element of the output is taken from the input instead of the passthru.  (If the passthru is undef, it doesn't matter.)